alternator/executor.cc

```diff
@@ -3573,6 +3573,11 @@ static rjson::value encode_paging_state(const schema& schema, const service::pag
             ++exploded_ck_it;
         }
     }
+    //FIXME conflicts with column names
```

Comment:
I consider this fixme a blocker for alternator.

As a first step, we should validate that it works on a subset of drivers (at least boto3 via our pytest suite), i.e. to make sure that the drivers don't do any validation, since you're going to return some column values as if they were keys, while they are not defined as such in the schema.


And then, if that works and we agree it's an acceptable divergence, we need to make sure there aren't any conflicts. Probably by reserving some magic column names unlikely to be ever used by anyone, in spirit of `:scylla:paging:region` and `:scylla:paging:weight`, unless somebody has a better suggestion.

Meanwhile, **if** it's trivial to skip alternator in this series and move it to a follow-up, perhaps it's the way to go, in case we're blocked on this discussion for a while. Perhaps the alternator code path could temporarily opt-out from getting a position "between" live rows, and just push for the first live one or eof?

Reply:
It's better to resolve this now IMO. This is a delicate issue and it's better to get it right rather than kick it down the road.

Reply:
I thought it would be a blocker. I don't think we can kick this down the road. Although the new feature is not used in this patchset yet and alternator could just opt out, later it would cause extra complexity (explicit flag to opt out) and alternator would keep suffering from workloads that other API will handle well.

I ran all the alternator tests with this: is that enough of a validation for the driver or is there some extra tests I should run?

Reply:
@nyh? I think that it's moderately safe to assume that the boto3 driver is Amazon's reference implementation, and it already contains lots of driver-side validation, so if the magic was accepted there, it's safe to treat this divergence as safe in compatibility terms. Personally I think that all drivers would treat this information similarly to how we manage our opaque paging state in CQL - just blindly copy it from the result into the next paged request, without any deep validation or changing anything.

Naturally, we also need to replace `region` and `weight` with custom, reserved names for the attributes, preferably ones containing reserved characters. Ref: https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/HowItWorks.NamingRulesDataTypes.html

Reply:
There is *some* validation, because my version had this JSON (example):
```json
{
    ...
    "weight": 1,
    "region": "clustered",
    ...
}
```
and the alternator driver did not like that one bit. I had to adjust these new fields to have the same value-object form that the other fields have. The driver doesn't seem to check the content though, only the form, i.e. the tests passed happily after I sorted out this issue and made sure I interpreted keys without weight/region as exclusive keys (like the "exclusiveSortKey" says).

Reply:
Technically, at least according to https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/HowItWorks.NamingRulesDataTypes.html:
```
Although DynamoDB allows you to use these reserved words and special characters for names, we recommend that you avoid doing so (...)
```
given that, in order to be 100% correct, perhaps we should validate early and check each attribute name to ensure that it isn't on our private reserved list, or, e.g., that it doesn't start with `:scylla:`, which we consider a reserved prefix. Otherwise it will be possible to craft a table which breaks this paging implementation. On the other hand, it's extremely unlikely that somebody would ever create a partition key named `:scylla:paging:weight`, so perhaps we can just let this one slip and not introduce needless checks on the fast path...

Reply:
(I'll unresolve this conversation for a short while until I get a response for my latest comment, feel free to reresolve it later)